pkg/controllers/proxy/utils.go

```diff
@@ -0,0 +1,158 @@
+		activeNetworkNamespaceHandle.String())
+	_ = activeNetworkNamespaceHandle.Close()
+
+	// TODO: fix boilerplate `netns.Set(hostNetworkNamespaceHandle)` code. Need a robust
+	// way to switch back to old namespace, pretty much all things will go wrong if we dont switch back
+
+	// create a ipip tunnel interface inside the endpoint container
+	tunIf, err := netlink.LinkByName(KubeTunnelIf)
+	if err != nil {
+		if err.Error() != IfaceNotFound {
+			attemptNamespaceResetAfterError(hostNetworkNamespaceHandle)
+			return fmt.Errorf("failed to verify if ipip tunnel interface exists in endpoint %s namespace due "+
+				"to %v", endpointIP, err)
+		}
+
+		klog.V(2).Infof("Could not find tunnel interface %s in endpoint %s so creating one.",
+			KubeTunnelIf, endpointIP)
+		ipTunLink := netlink.Iptun{
+			LinkAttrs: netlink.LinkAttrs{Name: KubeTunnelIf},
+			Local:     net.ParseIP(endpointIP),
+		}
+		err = netlink.LinkAdd(&ipTunLink)
+		if err != nil {
+			attemptNamespaceResetAfterError(hostNetworkNamespaceHandle)
+			return fmt.Errorf("failed to add ipip tunnel interface in endpoint namespace due to %v", err)
+		}
+
+		// this is ugly, but ran into issue multiple times where interface did not come up quickly.
+		for retry := 0; retry < 60; retry++ {
+			time.Sleep(interfaceWaitSleepTime)
+			tunIf, err = netlink.LinkByName(KubeTunnelIf)
+			if err == nil {
+				break
+			}
+			if err.Error() == IfaceNotFound {
+				klog.V(3).Infof("Waiting for tunnel interface %s to come up in the pod, retrying",
+					KubeTunnelIf)
+				continue
+			} else {
+				break
+			}
+		}
+
+		if err != nil {
+			attemptNamespaceResetAfterError(hostNetworkNamespaceHandle)
+			return fmt.Errorf("failed to get %s tunnel interface handle due to %v", KubeTunnelIf, err)
+		}
+
+		klog.V(2).Infof("Successfully created tunnel interface %s in endpoint %s.",
+			KubeTunnelIf, endpointIP)
+	}
+
+	// bring the tunnel interface up
+	err = netlink.LinkSetUp(tunIf)
+	if err != nil {
+		attemptNamespaceResetAfterError(hostNetworkNamespaceHandle)
+		return fmt.Errorf("failed to bring up ipip tunnel interface in endpoint namespace due to %v", err)
+	}
+
+	// assign VIP to the KUBE_TUNNEL_IF interface
+	err = ln.ipAddrAdd(tunIf, vip, false)
+	if err != nil && err.Error() != IfaceHasAddr {
+		attemptNamespaceResetAfterError(hostNetworkNamespaceHandle)
+		return fmt.Errorf("failed to assign vip %s to kube-tunnel-if interface", vip)
+	}
+	klog.Infof("Successfully assigned VIP: %s in endpoint %s.", vip, endpointIP)
+
+	// disable rp_filter on all interface
+	err = ioutil.WriteFile("/proc/sys/net/ipv4/conf/kube-tunnel-if/rp_filter",
+		[]byte(strconv.Itoa(0)), 0640)
+	if err != nil {
+		attemptNamespaceResetAfterError(hostNetworkNamespaceHandle)
+		return fmt.Errorf("failed to disable rp_filter on kube-tunnel-if in the endpoint container")
+	}
+
+	err = ioutil.WriteFile("/proc/sys/net/ipv4/conf/eth0/rp_filter", []byte(strconv.Itoa(0)), 0640)
```

Comment:
Is this assuming eth0 always exists?

Reply:
Should those paths be extracted into variables?

Reply:
Good catch on this one! That seems like it would be a long-standing bug.

Reply:
Actually, it turns out that all of this happens from inside the container. In this case the interface inside the container is always `eth0`. At this point, I'm not entirely sure what chooses this, but if I read the CNI spec correctly, it gets specified by the component that calls the CNI plugin to setup the device (https://www.cni.dev/docs/spec/#add-add-container-to-network-or-apply-modifications), which in the case of Kubernetes would be the `kubelet`.

However, I can't seem to find out what component of the kubelet chooses this name or how it is set. I've reached out to the SIG-NET Slack channel to see if they have an answer. After ~30 some min searching through the k8s code base, I can't seem to find it.

Reply:
I have added a TODO here to come back to this one. @thockin very helpfully provided in Slack that this is, in fact, determined by the container runtime. So the internal interface name itself could shift at some point in the future.

It looks like we can count on docker and containerd (https://github.com/containerd/containerd/blob/cf600abecc27200a3c0e1415cd1f6c325eb05edf/pkg/cri/server/helpers.go#L89) at least to name this eth0 for now, but this could shift at any point in the future or with a different runtime.